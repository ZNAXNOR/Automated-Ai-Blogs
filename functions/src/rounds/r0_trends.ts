/* 
 * Deterministic, zero-cost by default. Optionally prunes/scores via a tiny HF model.
 *
 * Inputs:
 *   runId: string
 *   seeds: string[]
 *   region?: string
 *   useLLM?: boolean (default false)
 *
 * Side effects:
 *   - Writes artifact to Firestore: runs/{runId}.artifacts.round0
 *   - Caches SerpApi responses under cache/serpapi/{hash} with TTL
 *
 * Notes:
 *   - If SerpApi key not present, we fall back to RSS-only path.
 *   - The deterministic path is always executed; the LLM step is an optional extra prune.
 */

import * as admin from "firebase-admin";
import { HttpsError } from "firebase-functions/v2/https";
import { onCall } from "firebase-functions/v2/https";
import { FieldValue } from "firebase-admin/firestore";
import { env } from "../utils/config"; // expects SERP_API_KEY, HF_TOKEN, USE_R0_LLM, CACHE_TTL_HOURS
import { getSerpSuggestions, getSerpRelated, getSerpTrending, serpAvailable } from "../clients/serp";
import { getCache, setCache } from "../utils/cache";
import * as crypto from "crypto";

// --- Types -------------------------------------------------------------------

export interface TrendItem {
  query: string; // normalized short query
  type: "autocomplete" | "related" | "trending" | "rss";
  score: number; // 0..1
  source: string[]; // e.g., ["serp:autocomplete","rss:theverge"]
  reason?: string; // <= 12 words
}

interface Round0Input {
  runId: string;
  seeds: string[];
  region?: string;
  useLLM?: boolean;
  force?: boolean;
}

type SourceBucket = {
  type: TrendItem["type"];
  sourceName: string; // "serp:autocomplete" | "serp:related" | "serp:trending" | "rss:<domain>"
  items: string[]; // raw candidate strings (unnormalized)
};

// --- Admin init (safe if already initialized) --------------------------------
if (!admin.apps.length) {
  admin.initializeApp();
}
const db = admin.firestore();

// --- Lightweight RSS sources (fallback) --------------------------------------
// You can safely modify/extend this list.
const RSS_SOURCES: { name: string; url: string }[] = [
  { name: "rss:theverge", url: "https://www.theverge.com/rss/index.xml" },
  { name: "rss:techcrunch", url: "https://techcrunch.com/feed/" },
  { name: "rss:bbcworld", url: "https://feeds.bbci.co.uk/news/world/rss.xml" },
  { name: "rss:reutersworld", url: "https://feeds.reuters.com/reuters/worldNews" },
];

// --- Utilities ---------------------------------------------------------------

const ROUND = 0;

function logStep(runId: string, step: string, startMs: number) {
  const durationMs = Date.now() - startMs;
  // Use console.log for structured logs in emulator & prod
  console.log(JSON.stringify({ runId, round: ROUND, step, durationMs }));
}

function assertInput(i: any): asserts i is Round0Input {
  if (!i || typeof i !== "object") throw new HttpsError("invalid-argument", "Input must be an object");
  if (!i.runId || typeof i.runId !== "string") throw new HttpsError("invalid-argument", "runId is required");
  if (!Array.isArray(i.seeds) || i.seeds.some((s: string) => typeof s !== "string")) {
    throw new HttpsError("invalid-argument", "seeds must be a string[]");
  }
  if (i.region && typeof i.region !== "string") {
    throw new HttpsError("invalid-argument", "region must be a string");
  }
  if (i.useLLM !== undefined && typeof i.useLLM !== "boolean") {
    throw new HttpsError("invalid-argument", "useLLM must be boolean");
  }
}

function sha256(str: string) {
  return crypto.createHash("sha256").update(str, "utf8").digest("hex");
}

// Keep '+', '#', and '-'; strip other punctuation.
function stripPunctKeepSymbols(s: string): string {
  return s.replace(/[!"$%&'()*.,\/:;<=>?@\[\\\]^_`{|}~]/g, " "); // keep +, #, and -
}

// Normalize: lowercase, trim, collapse spaces, truncate to 6 words
export function normalizeQuery(raw: string): string {
  const s = stripPunctKeepSymbols(raw.toLowerCase())
    .replace(/\s+/g, " ")
    .trim();
  const words = s.split(" ").filter(Boolean).slice(0, 6);
  return words.join(" ");
}

// Filters per spec
export function shouldDrop(q: string): boolean {
  const tokens = q.split(" ").filter(Boolean);
  const hasPersonal = /\b(my|me|account|password)\b/.test(q);
  if (hasPersonal) return true;

  const numericCount = tokens.filter(t => /^\d+([\/\.\-]\d+)*$/.test(t)).length;
  if (tokens.length > 0 && numericCount / tokens.length > 0.6) return true;

  const generic = ["news", "update", "latest"];
  const containsGeneric = tokens.some(t => generic.includes(t));
  if (containsGeneric) {
    // allow if paired with context like "apple latest" (â‰¥ 2 tokens and not only generic)
    if (tokens.length === 1 && generic.includes(tokens[0])) return true;
    const nonGenericCount = tokens.filter(t => !generic.includes(t)).length;
    if (nonGenericCount === 0) return true;
  }
  return false;
}

// Simple Levenshtein (enough for small N)
function levenshtein(a: string, b: string): number {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;
  const v0 = Array(b.length + 1).fill(0);
  const v1 = Array(b.length + 1).fill(0);
  for (let i = 0; i <= b.length; i++) v0[i] = i;
  for (let i = 0; i < a.length; i++) {
    v1[0] = i + 1;
    for (let j = 0; j < b.length; j++) {
      const cost = a[i] === b[j] ? 0 : 1;
      v1[j + 1] = Math.min(v1[j] + 1, v0[j + 1] + 1, v0[j] + cost);
    }
    for (let j = 0; j <= b.length; j++) v0[j] = v1[j];
  }
  return v1[b.length];
}

function tokenOverlap(a: string, b: string): number {
  const A = new Set(a.split(" ").filter(Boolean));
  const B = new Set(b.split(" ").filter(Boolean));
  if (A.size === 0 && B.size === 0) return 1;
  let inter = 0;
  A.forEach(t => {
    if (B.has(t)) inter++;
  });
  const union = new Set([...A, ...B]).size;
  return inter / union;
}

function nearDuplicate(a: string, b: string): boolean {
  return levenshtein(a, b) <= 2 || tokenOverlap(a, b) >= 0.75;
}

// Optionally call a tiny HF LLM to prune/score (kept very cheap)
// If env.HF_TOKEN missing or useLLM=false, skip and return passthrough.
async function optionalLlmPrune(
  items: TrendItem[],
  useLLM: boolean | undefined
): Promise<TrendItem[]> {
  if (!useLLM && !env.useR0Llm) return items;
  if (!env.HF_TOKEN) return items;

  // Token-efficient prompt: rank by general audience interest + freshness proxies
  const prompt =
`You are scoring short search queries for blog topics.
Keep 12 or fewer. Prefer widely interesting, multi-source, and non-duplicative.
Return as CSV: query,boost where boost in [0..0.2].

Items:
${items.map(i => `- ${i.query} [${i.type}]`).join("\n")}
`;

  try {
    // Lazy import to avoid dependency if unused
    const { hfTinyComplete } = await import("../clients/hf");
    const csv = await hfTinyComplete(prompt); // returns small CSV string or empty
    if (!csv) return items;

    const boosts = new Map<string, number>();
    csv.split(/\r?\n/).forEach(line => {
      const m = line.split(",");
      if (m.length >= 2) {
        const q = normalizeQuery(m[0] || "");
        const b = Number(m[1]);
        if (q && isFinite(b)) boosts.set(q, Math.max(0, Math.min(0.2, b)));
      }
    });

    const boosted = items.map(i => ({
      ...i,
      score: Math.max(0, Math.min(1, i.score + (boosts.get(i.query) ?? 0))),
    }));

    // Re-sort and clamp to 12
    boosted.sort((a, b) => (b.score - a.score) || (a.query.length - b.query.length));
    return boosted.slice(0, 12);
  } catch (e) {
    console.warn("LLM prune skipped due to error:", e);
    return items;
  }
}

// --- RSS fetch & parse (lightweight) ----------------------------------------
// Minimal parser: extract <title>...</title> from top items (ignores CDATA nuances gracefully)
async function fetchRssTitles(url: string, limit = 20): Promise<string[]> {
  const res = await fetch(url, { method: "GET", redirect: "follow" });
  const text = await res.text();
  const titles: string[] = [];
  const re = /<title>([\s\S]*?)<\/title>/gi;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const raw = m[1]
      .replace(/<!\[CDATA\[(.*?)\]\]>/g, "$1")
      .replace(/\s+/g, " ")
      .trim();
    // Skip the feed-level title (usually first one)
    if (raw && titles.length < limit) titles.push(raw);
  }
  // Remove the first title if it is the channel title (heuristic)
  return titles.length > 1 ? titles.slice(1) : titles;
}

// --- Core deterministic pipeline --------------------------------------------

export function deterministicProcess(
  buckets: SourceBucket[]
): { items: TrendItem[]; sourceCounts: Record<string, number> } {
  const collected = new Map<string, TrendItem>();

  for (const b of buckets) {
    for (const raw of b.items) {
      const q = normalizeQuery(raw);
      if (!q || shouldDrop(q)) continue;

      if (collected.has(q)) {
        // Merge source/type
        const existing = collected.get(q)!;
        const newSources = new Set([...existing.source, b.sourceName]);
        existing.source = [...newSources];
        const typeRank = (t: TrendItem["type"]) =>
          t === "trending" ? 3 : t === "related" ? 2 : t === "autocomplete" ? 1 : 0;
        if (typeRank(b.type) > typeRank(existing.type)) {
          existing.type = b.type;
        }
      } else {
        collected.set(q, {
          query: q,
          type: b.type,
          score: 0,
          source: [b.sourceName],
        });
      }
    }
  }

  // Near-duplicate merge
  const merged: TrendItem[] = [];
  for (const item of Array.from(collected.values())) {
    const dupIdx = merged.findIndex(m => nearDuplicate(m.query, item.query));
    if (dupIdx === -1) {
      merged.push(item);
    } else {
      // merge sources and keep best type (prefer trending > related > autocomplete > rss)
      const target = merged[dupIdx];
      const src = new Set([...target.source, ...item.source]);
      target.source = [...src];
      const typeRank = (t: TrendItem["type"]) =>
        t === "trending" ? 3 : t === "related" ? 2 : t === "autocomplete" ? 1 : 0;
      if (typeRank(item.type) > typeRank(target.type)) target.type = item.type;
    }
  }

  // Scoring
  const fromTrending = new Set(
    merged.filter(i => i.source.some(s => s.startsWith("serp:trending"))).map(i => i.query)
  );

  const itemsScored = merged.map(i => {
    let score = 0.5; // base
    const multiSource = i.source.length > 1;
    if (multiSource) score += 0.10;
    if (fromTrending.has(i.query)) score += 0.10;
    score = Math.max(0, Math.min(1, score));
    const reasonBits = [];
    if (fromTrending.has(i.query)) reasonBits.push("trending");
    if (multiSource) reasonBits.push("multi-source");
    const reason = reasonBits.length ? reasonBits.join(", ").slice(0, 48) : undefined;
    return { ...i, score, reason };
  });

  // Sort by score desc then by query length asc; clamp 12
  itemsScored.sort((a, b) => (b.score - a.score) || (a.query.length - b.query.length));
  const final = itemsScored.slice(0, 12);

  // sourceCounts
  const sourceCounts: Record<string, number> = {};
  for (const i of final) {
    for (const s of i.source) sourceCounts[s] = (sourceCounts[s] ?? 0) + 1;
  }

  return { items: final, sourceCounts };
}

// --- Firestore artifact helpers ---------------------------------------------

async function getExistingArtifact(runId: string) {
  const ref = db.doc(`runs/${runId}`);
  const snap = await ref.get();
  const data = snap.exists ? snap.data() : undefined;
  const art = data?.artifacts?.round0;
  return art ? { ...art } : null;
}

async function writeArtifact(runId: string, payload: any) {
  const ref = db.doc(`runs/${runId}`);
  await ref.set(
    {
      artifacts: {
        round0: {
          ...payload,
          createdAt: FieldValue.serverTimestamp(),
        },
      },
    },
    { merge: true }
  );
}

// --- Main callable -----------------------------------------------------------

export const round0_trends = onCall(
  { timeoutSeconds: 300, memory: "256MiB" },
  async (req) => {
    const t0 = Date.now();
    const input = req.data;
    assertInput(input);
    const { runId, seeds, region, useLLM, force } = input;

    // Idempotency
    if (!force) {
      const existing = await getExistingArtifact(runId);
      if (existing) {
        logStep(runId, "idempotent-return", t0);
        return existing;
      }
    }

    const tSerp = Date.now();

    // --- Gather sources (Serp + RSS fallback) --------------------------------
    const buckets: SourceBucket[] = [];

    // SerpApi (if available)
    let usedSerp = false;
    let serpCacheHit = false;

    if (serpAvailable()) {
      usedSerp = true;
      const dayBucket = new Date().toISOString().slice(0, 10); // YYYY-MM-DD for cache key
      const baseKey = JSON.stringify({ seeds, region, day: dayBucket });

      // Autocomplete
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":autocomplete")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpSuggestions(seeds, region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "autocomplete", sourceName: "serp:autocomplete", items: list });
      }
      // Related
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":related")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpRelated(seeds, region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "related", sourceName: "serp:related", items: list });
      }
      // Trending
      {
        const cacheKey = `serpapi:${sha256(baseKey + ":trending")}`;
        const cached = await getCache(cacheKey);
        let list: string[];
        if (cached?.payload) {
          list = cached.payload;
          serpCacheHit = true;
        } else {
          list = await getSerpTrending(region);
          await setCache(cacheKey, list, env.CACHE_TTL_HOURS);
        }
        buckets.push({ type: "trending", sourceName: "serp:trending", items: list });
      }
    }

    // RSS fallback (always included, but especially useful if Serp is unavailable)
    const tRss = Date.now();
    for (const src of RSS_SOURCES) {
      try {
        const titles = await fetchRssTitles(src.url, 20);
        buckets.push({ type: "rss", sourceName: src.name, items: titles });
      } catch (e) {
        console.warn(`RSS fetch error for ${src.name}`, e);
      }
    }
    logStep(runId, "fetch-sources", tSerp);

    // --- Deterministic normalization/dedup/score ------------------------------
    const tDet = Date.now();
    const { items: detItems, sourceCounts } = deterministicProcess(buckets);
    logStep(runId, "deterministic", tDet);

    // --- Optional LLM prune/boost --------------------------------------------
    const tLlm = Date.now();
    const finalItems = await optionalLlmPrune(detItems, useLLM);
    logStep(runId, "optional-llm", tLlm);

    // --- Output validation (simple) ------------------------------------------
    for (const i of finalItems) {
      if (!i.query || typeof i.query !== "string") throw new HttpsError("internal", "Invalid item.query");
      if (!["autocomplete", "related", "trending", "rss"].includes(i.type)) {
        throw new HttpsError("internal", "Invalid item.type");
      }
      if (typeof i.score !== "number" || i.score < 0 || i.score > 1) {
        throw new HttpsError("internal", "Invalid item.score");
      }
      if (!Array.isArray(i.source)) throw new HttpsError("internal", "Invalid item.source");
      if (i.reason && i.reason.split(/\s+/).length > 12) {
        throw new HttpsError("internal", "Invalid item.reason length");
      }
    }

    // --- Write artifact -------------------------------------------------------
    const artifact = {
      items: finalItems,
      cached: serpCacheHit,
      sourceCounts,
    };
    await writeArtifact(runId, artifact);

    logStep(runId, "done", t0);
    return artifact;
  }
);

// Export internals for unit testing
export const _test = {
  normalizeQuery,
  shouldDrop,
  deterministicProcess,
  nearDuplicate,
};
