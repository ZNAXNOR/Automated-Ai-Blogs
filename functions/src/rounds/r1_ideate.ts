import admin from 'firebase-admin';
import fetch from 'node-fetch';
import { z } from 'zod';
import { env } from '../utils/config';
import { logger } from '../utils/logger';
import { ResponseWrapper } from '../utils/responseHelper';
import { ARTIFACT_PATHS } from '../utils/constants';
import {
  Round1InputSchema,
  Round1OutputSchema,
} from '../utils/schema';

export type IdeationItem = z.infer<typeof Round1OutputSchema>["items"][number];

const MAX_IDEAS_PER_TREND = 5;
const MIN_IDEAS_PER_TREND = 3;
const MAX_TOTAL_IDEAS = 60;

if (!admin.apps.length) {
  admin.initializeApp();
}

// Schema for the raw output from the LLM
const LlmResponseSchema = z.array(
  z.object({
    trend: z.string(),
    ideas: z.array(z.string()),
  })
);

function buildPrompt(trendQueries: string[]): string {
  const prompt = `You are an expert blog strategist helping select article titles from trending topics.

TASK:
- For each input trend, generate 3–5 unique, creative blog title ideas.
- Titles should be short (≤12 words), clear, and engaging.
- Focus on human readability and search-friendliness.
- Avoid clickbait and repetition.

OUTPUT FORMAT (strict JSON only):
[
  {
    "trend": "<trend string>",
    "ideas": ["<title1>", "<title2>", "<title3>"]
  },
  ...
]

EXAMPLE:
Input: ["AI in healthcare"]
Output:
[
  {
    "trend": "AI in healthcare",
    "ideas": [
      "How AI is Transforming Healthcare in 2025",
      "AI in Hospitals: Benefits and Challenges",
      "The Future of Medicine with Artificial Intelligence"
    ]
  }
]

Input: ${JSON.stringify(trendQueries)}
`;
  return prompt;
}

async function callHuggingFace(prompt: string): Promise<ResponseWrapper> {
  const hfToken = env.hfToken;
  const HF_MODEL = env.hfModelR1;

  if (!hfToken) {
    throw new Error('HF_TOKEN environment variable is not set.');
  }
  if (!HF_MODEL || HF_MODEL.includes('<set-your-model')) {
    throw new Error(
      'HUGGINGFACE_MODEL environment variable is not set to a valid model slug.'
    );
  }

  const HF_ENDPOINT = `https://api-inference.huggingface.co/models/${HF_MODEL}`;

  const payload = {
    inputs: prompt,
    parameters: {
      max_new_tokens: 512,
      return_full_text: true,
    },
  };

  const res = await fetch(HF_ENDPOINT, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${hfToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const text = await res.text();
    logger.error('Hugging Face API error', { status: res.status, text });
    throw new Error(`Hugging Face API error: ${res.status} ${res.statusText} - ${text}`);
  }

  return ResponseWrapper.create(res);
}

function mapToIdeationItems(parsed: z.infer<typeof LlmResponseSchema>): IdeationItem[] {
  const items: IdeationItem[] = [];
  for (const entry of parsed) {
    const trend = String(entry.trend ?? '').trim();
    const ideas = Array.isArray(entry.ideas) ? entry.ideas : [];

    if (!trend) continue;
    const cleanedIdeas = ideas
      .map((s: any) => String(s ?? '').trim())
      .filter((s: string) => s.length > 0)
      .slice(0, MAX_IDEAS_PER_TREND);

    if (cleanedIdeas.length < MIN_IDEAS_PER_TREND) {
      continue;
    }

    cleanedIdeas.forEach((idea: string, idx: number) => {
      items.push({
        trend,
        idea,
        variant: idx + 1,
        source: 'llm',
      });
    });
  }

  if (items.length > MAX_TOTAL_IDEAS) {
    items.splice(MAX_TOTAL_IDEAS);
  }

  return items.filter((it) => it.idea.trim().length > 0);
}

export async function Round1_Ideate(runId: string): Promise<{ wrote: number }> {
  logger.info('Round 1: Ideate starting', { runId });
  if (!runId) throw new Error('runId is required');

  const db = admin.firestore();

  const r0DocRef = db.collection('runs').doc(runId).collection('artifacts').doc('round0');
  const r0Snap = await r0DocRef.get();
  if (!r0Snap.exists) {
    logger.error('Round0 artifact not found', { runId });
    throw new Error(`Round0 artifact not found for runId=${runId} at runs/${runId}/artifacts/round0`);
  }

  const r0Data = r0Snap.data() || {};

  const validationResult = Round1InputSchema.safeParse(r0Data);
  if (!validationResult.success) {
    logger.error('Round 1 Input validation failed', {
      runId,
      error: validationResult.error,
    });
    throw new Error('Round 1 Input validation failed');
  }

  const { items: trendsFromDoc } = validationResult.data;

  if (!Array.isArray(trendsFromDoc) || trendsFromDoc.length === 0) {
    throw new Error(`No trends found in round0 artifact for runId=${runId}`);
  }

  const trendQueries = Array.from(
    new Set(trendsFromDoc.map((t) => String(t.query ?? '').trim()).filter((s) => s.length > 0))
  ).slice(0, 12);

  if (trendQueries.length === 0) {
    throw new Error('After normalization, no valid trend queries were found.');
  }

  const prompt = buildPrompt(trendQueries);

  const hfResponse = await callHuggingFace(prompt);
  const parsed = await hfResponse.json(LlmResponseSchema);

  const ideationItems = mapToIdeationItems(parsed);

  if (ideationItems.length === 0) {
    throw new Error('No valid ideation items produced by LLM.');
  }

  const trendsMap = new Map<string, number>();
  for (const it of ideationItems) {
    trendsMap.set(it.trend, (trendsMap.get(it.trend) || 0) + 1);
  }

  for (const t of trendQueries) {
    const count = trendsMap.get(t) || 0;
    if (count < MIN_IDEAS_PER_TREND) {
      throw new Error(`Trend \"${t}\" produced fewer than ${MIN_IDEAS_PER_TREND} ideas (${count}).`);
    }
  }

  const outputForValidation: z.infer<typeof Round1OutputSchema> = {
    items: ideationItems,
  };
  const outputValidationResult = Round1OutputSchema.safeParse(outputForValidation);

  if (!outputValidationResult.success) {
    logger.error('Round 1 Output validation failed', {
      runId,
      error: outputValidationResult.error,
    });
    throw new Error('Round 1 Output validation failed');
  }

  const r1ArtifactPath = ARTIFACT_PATHS.R1_IDEATION.replace('{runId}', runId);
  const r1DocRef = db.doc(r1ArtifactPath);

  await r1DocRef.set({
    createdAt: admin.firestore.FieldValue.serverTimestamp(),
    ...outputValidationResult.data,
  });

  logger.info('Round 1: Ideate finished', { runId, wrote: ideationItems.length });
  return { wrote: ideationItems.length };
}
