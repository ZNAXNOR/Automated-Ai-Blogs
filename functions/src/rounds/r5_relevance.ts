/**
 * - Takes Round 4 polished text and derivatives
 * - Assesses relevance of the polished text against the original draft
 * - Stores results in Firestore collection: round5_relevance
 */

import { onCall } from "firebase-functions/v2/https";
import { getFirestore } from "firebase-admin/firestore";
import { env } from "../utils/config";
import fetch from "node-fetch";

function buildPrompt(originalDraft: string, polishedText: string): string {
  return `
System: You are a content marketing analyst. Your task is to evaluate whether a polished text, derived from an original draft, remains true to the core message and intent of the original. 
If the polished text is a relevant and faithful adaptation, respond with {"isRelevant": true, "reason": "The polished text successfully captures the essence of the original draft."}. 
If it deviates significantly, respond with {"isRelevant": false, "reason": "[Explain the specific discrepancy]".

User: Here is the original draft and the polished text.

Original Draft:
${originalDraft}

Polished Text:
${polishedText}
`;
}

async function callHuggingFace(prompt: string): Promise<{ isRelevant: boolean; reason: string }> {
    const HF_API_URL = `https://api-inference.huggingface.co/models/${env.hfModelR5}`;

    const response = await fetch(HF_API_URL, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${env.HF_API_KEY}`,
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            inputs: prompt,
            parameters: {
                max_new_tokens: 256,
                temperature: 0.5,
                return_full_text: false,
            },
        }),
    });

    if (!response.ok) {
        throw new Error(`HF API error: ${response.status} ${response.statusText}`);
    }

    const data = (await response.json()) as any;

    let text: string;
    if (Array.isArray(data) && data.length > 0 && data[0].generated_text) {
        text = data[0].generated_text;
    } else if (typeof data === "object" && data.generated_text) {
        text = data.generated_text;
    } else {
        text = JSON.stringify(data);
    }

    try {
        const jsonMatch = text.match(/\{[\s\S]*\}/);
        if (!jsonMatch) {
            console.error("No JSON found in model response:", text);
            throw new Error("Failed to parse JSON from model: No JSON found");
        }
        const parsed = JSON.parse(jsonMatch[0]);
        return parsed;
    } catch (err) {
        console.error("Failed to parse JSON from model:", text);
        throw err;
    }
}

async function runR5Relevance(draftId: string, originalDraft: string, polishedText: string) {
  if (!draftId || !originalDraft || !polishedText) {
    throw new Error("Missing required fields: draftId, originalDraft, polishedText");
  }
  
  const db = getFirestore();

  const prompt = buildPrompt(originalDraft, polishedText);

  // Call LLM
  const { isRelevant, reason } = await callHuggingFace(prompt);

  // Save to Firestore
  const docRef = db.collection("round5_relevance").doc(draftId);
  await docRef.set({
    draftId,
    isRelevant,
    reason,
    createdAt: new Date().toISOString(),
  });

  return { draftId, isRelevant, reason };
}

export const r5_relevance = onCall(
  { timeoutSeconds: 300, memory: "1GiB" },
  async (request) => {
    const { draftId, originalDraft, polishedText } = request.data;
    return runR5Relevance(draftId, originalDraft, polishedText);
  }
);

// Export for testing
export const _test = {
    buildPrompt,
    runR5Relevance
};
