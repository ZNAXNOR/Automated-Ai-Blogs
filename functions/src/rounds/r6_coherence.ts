import * as admin from "firebase-admin";
import { onCall, HttpsError } from "firebase-functions/v2/https";
import { z } from "zod";
import pLimit from "p-limit";
import { env } from "../utils/config";
import { logger } from "../utils/logger";
import { constants } from "../utils/constants";
import { calculateSimilarity } from "../clients/hf_sentence";
import { R5Meta, R6Coherence, JobPayload } from "../utils/types";

// --- Schemas ------------------------------------------------------------------

const R5OutputSchema = z.object({
  items: z.array(R5Meta),
});

const FailedItemSchema = z.object({
  item: R5Meta,
  error: z.string(),
});

// --- Constants ----------------------------------------------------------------

const ROUND = 6;
const CONCURRENCY = 5;

// --- Admin SDK ----------------------------------------------------------------

if (!admin.apps.length) {
  admin.initializeApp();
}
const db = admin.firestore();

// --- Helper Functions ---------------------------------------------------------

async function getR5Data(runId: string): Promise<z.infer<typeof R5OutputSchema>> {
  const docRef = db.doc(constants.ARTIFACT_PATHS.R5_METADATA.replace("{runId}", runId));
  const docSnap = await docRef.get();
  if (!docSnap.exists) {
    throw new HttpsError("not-found", `Round 5 artifact not found for runId=${runId}`);
  }

  const validationResult = R5OutputSchema.safeParse(docSnap.data());
  if (!validationResult.success) {
    logger.error("Round 5 data validation failed", { runId, error: validationResult.error });
    throw new HttpsError("internal", "Round 5 data validation failed");
  }
  if (validationResult.data.items.length === 0) {
    throw new HttpsError("failed-precondition", "R5 artifact has no items.");
  }

  return validationResult.data;
}

async function analyzeCoherence(
  item: R5Meta
): Promise<R6Coherence> {
  const sentences = item.draft.split(". ");
  const scores = await calculateSimilarity(item.draft, sentences);
  const averageScore = scores.reduce((sum, score) => sum + score, 0) / scores.length;

  return {
    ...item,
    coherenceScores: {
      overall: averageScore,
      sentence: scores,
    },
  };
}

async function writeArtifacts(
  runId: string,
  successfulItems: R6Coherence[],
  failedItems: z.infer<typeof FailedItemSchema>[]
): Promise<void> {
  if (successfulItems.length === 0 && failedItems.length === 0) {
    logger.warn("No artifacts to write for Round 6.", { runId });
    return;
  }

  const batch = db.batch();

  if (successfulItems.length > 0) {
    const successPath = constants.ARTIFACT_PATHS.R6_COHERENCE.replace("{runId}", runId);
    batch.set(db.doc(successPath), {
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
      items: successfulItems,
    });
  }

  // No explicit failure artifact for R6 yet, just logging.

  await batch.commit();
}

// --- Main Function ------------------------------------------------------------

export async function run(
  payload: JobPayload
): Promise<{ coherenceCount: number; failures: number }> {
  const { runId } = payload;
  if (typeof runId !== "string" || !runId) {
    throw new HttpsError("invalid-argument", "runId must be a non-empty string.");
  }
  logger.info(`Round ${ROUND}: Coherence starting`, { runId });

  const { items: r5Items } = await getR5Data(runId);

  const limit = pLimit(CONCURRENCY);
  const successfulItems: R6Coherence[] = [];
  const failedItems: z.infer<typeof FailedItemSchema>[] = [];

  const promises = r5Items.map((item) =>
    limit(async () => {
      try {
        const coherenceItem = await analyzeCoherence(item);
        successfulItems.push(coherenceItem);
      } catch (error: any) {
        logger.error(`Failed to analyze coherence for \"${item.title}\"`, { runId, error: error.message });
        failedItems.push({ item, error: error.message });
      }
    })
  );

  await Promise.all(promises);

  await writeArtifacts(runId, successfulItems, failedItems);

  const result = { coherenceCount: successfulItems.length, failures: failedItems.length };
  logger.info(`Round ${ROUND}: Coherence finished`, { runId, ...result });
  return result;
}

export const Round6_Coherence = onCall(
  { timeoutSeconds: 180, memory: "256MiB", region: env.region },
  (req) => run(req.data)
);
