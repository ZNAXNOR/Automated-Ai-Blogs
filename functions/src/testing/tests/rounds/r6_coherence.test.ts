jest.mock("node-fetch", () => jest.fn());

import { Round6_Coherence } from "../../../rounds/r6_coherence";

import fetch from "node-fetch";
const mockedFetch = fetch as unknown as jest.Mock;

// Mock environment variables
jest.mock("../../../utils/config", () => ({
  env: {
    hfToken: "test-token",
    hfModelR6: "test-model",
  },
}));


// --- Mocking Firestore -- -
const firestoreWrites: { [key: string]: any } = {};
const collectionGetMock = jest.fn().mockResolvedValue({
    empty: false,
    docs: [
        {
            id: 'draft1',
            data: () => ({
                polished: 'This is the first polished draft.',
                derivatives: ['derivative A', 'derivative B'],
            }),
        },
        {
            id: 'draft2',
            data: () => ({
                polished: 'This is the second polished draft.',
                derivatives: ['derivative C', 'derivative D'],
            }),
        },
    ],
});

const batchSetMock = jest.fn((docRef, data) => {
    firestoreWrites[docRef.path] = data;
});
const batchCommitMock = jest.fn().mockResolvedValue(undefined);

const collectionMock = jest.fn((path: string) => {
    if (path.includes("round4_distribution")) {
        return {
            get: collectionGetMock
        };
    }
    return {
        doc: (docId: string) => ({
            path: `${path}/${docId}`,
        }),
    }
});

jest.mock("firebase-admin/firestore", () => ({
    getFirestore: () => ({
        collection: collectionMock,
        batch: () => ({
            set: batchSetMock,
            commit: batchCommitMock,
        }),
    }),
    FieldValue: {
        serverTimestamp: () => 'mock-server-timestamp',
    },
}));


describe("r6_coherence Core Functionality & Firestore Writes", () => {

    beforeEach(() => {
        jest.clearAllMocks();
        mockedFetch.mockReset();
        for (const key in firestoreWrites) {
            delete firestoreWrites[key];
        }
    });

    it("should calculate coherence, save to Firestore, and match snapshot", async () => {
        mockedFetch.mockResolvedValue({
            ok: true,
            json: async () => [0.9, 0.95],
        });

        const runId = "test-run-for-snapshot";
        await Round6_Coherence(runId);

        expect(collectionMock).toHaveBeenCalledWith(`runs/${runId}/artifacts/round4_distribution`);
        expect(fetch).toHaveBeenCalledTimes(2);
        expect(batchCommitMock).toHaveBeenCalledTimes(1);

        // Sort keys to ensure consistent snapshot
        const sortedWrites = Object.keys(firestoreWrites).sort().reduce(
            (acc, key) => {
                acc[key] = firestoreWrites[key];
                return acc;
            }, {} as { [key: string]: any }
        );

        expect(sortedWrites).toMatchSnapshot();
    });
});
