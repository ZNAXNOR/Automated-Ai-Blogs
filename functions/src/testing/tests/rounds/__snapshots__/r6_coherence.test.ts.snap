// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`r6_coherence Core Functionality & Firestore Writes should calculate coherence, save to Firestore, and match snapshot 1`] = `
{
  "runs/test-run-for-snapshot/artifacts/round6/draft1": {
    "coherenceScore": 0.925,
    "createdAt": "mock-server-timestamp",
    "derivatives": [
      "derivative A",
      "derivative B",
    ],
    "draftId": "draft1",
    "validatedText": "This is the first polished draft.",
  },
  "runs/test-run-for-snapshot/artifacts/round6/draft2": {
    "coherenceScore": 0.925,
    "createdAt": "mock-server-timestamp",
    "derivatives": [
      "derivative C",
      "derivative D",
    ],
    "draftId": "draft2",
    "validatedText": "This is the second polished draft.",
  },
}
`;
